### DISCUSSION TITLE
Using pyFirmata for Peeps with Arduinos

### DISCUSSION THREAD

**Discussion by jmdelahanty** [wehr-lab/autopilot#180] [Ideas]
Hey everyone! People in the lab are quite fond of them and seem to trust the timings/clock consistency of them more than the pi. A talented programmer in our lab, Annie (not sure what her github handle is yet...), mentioned that making a plugin for Autopilot to run with Arduinos would be cool.

While looking into how to communicate with Arduinos/send sketches over the wire to them in Python, I came across a library called [firmata](https://github.com/firmata/arduino) and a Python library called [pyFirmata](https://github.com/tino/pyFirmata). Maybe these libraries could be used to generate an interface with Arduinos in Autopilot? Just figured it would be cool to have this on the discussions forum as well.

**Comment by sneakers-the-rat** [comment]
i have only heard this timing precision argument literally every time someone asks about the pi, but it's an empirical question! in my experience it's mostly superstition and in my opinion optimizing for the wrong thing when weight against the complexity and maintainability of the code. But define what kind of accuracy you need and try measuring it! here's the jitter we get in a roundtrip external input -> read -> write test, so it includes both read and write. The read also happens with hardware interrupts and pigpio passes in the CPU tick from when the logic level changes, so idk i think it's pretty accurate. test details and code to replicate is in the paper.

<img width="872" alt="Screen Shot 2022-07-14 at 8 03 20 PM" src="https://user-images.githubusercontent.com/12961499/179141704-3e70c82a-bd19-402c-a57c-361740ff0974.png">

so if you need better than 50 microseconds IQR jitter then by all means! 

Avoiding needing to write a virtualization wrapper to a microcontroller is one of the main reasons to use the pi, but if that's the route you want to go then that would be totally doable, that's basically what bpod does. You'd basically just need to define what you would want from such an interface and then i can help coach through - eg. what kind of instructions do you want to be able to send and receive from the microcontroller? Using the pi to do things like individual GPIO flips on the arduino is sort of pointless, as you then just add whatever jitter/latency from routing through serial/etc. to the jitter/latency you were trying to avoid from the PI. But otherwise it's totally possible to write a firmware for the microcontroller that can accept instructions from the pi, would just need to know what!

**Comment by cxrodgers** [comment]
Before I started using Autopilot, I wrote a behavioral control system entirely using Arduino. IMO the biggest problem is that debugging on Arduino is much more difficult -- no log files, no ability to ssh in or hook up a display, all you can do is write messages to the serial port ... but in a crash scenario, that probably won't work anyway. As a result I found that code of even moderate complexity became very brittle on Arduino. 

I do like using Arduino for fast and simple tasks -- for instance, Autopilot can raise a digital line, and this triggers the Arduino to execute a precisely timed sequence to flash LEDs. I think of the Arduino as a bit like the "spinal cord" in this scenario. But once things get complicated, I think you're way better off on the Pi.

**Comment by jmdelahanty** [comment]
As far as I'm aware, we definitely don't need better than 50us IQR jitter for things! The main reason I posted this was because it's one of the most common threads about adopting Autopilot that I've run into so far. I've tried talking about how you have indeed benchmarked these things and that the Pi, especially when used how you designed it with the cool multi-threading of things, is likely plenty fast! When looking around at ways to assuage those fears, beyond the ones in your paper, I found this stuff and thought it could be interesting.

I personally would very much like to use the pi for our headfixed discrimination task (similar to 2AFC, but even simpler than that lol) but haven't had much success in getting fellow lab mates to join on the adventure yet...

>  As a result I found that code of even moderate complexity became very brittle on Arduino.

This is something that Jonny pointed out to me about how we've been doing things on Arduinos and I can't stop thinking about it! The elegance of how Autopilot can make things so simple but also so robust is just something that makes me smile. Also after reading the Autopilot paper again over the weekend, I started trying to understand the notion of how Arduinos act as state machines and how limiting that can be for behavior!

On the speed side of things, I'm a little confused about how sampling is actually performed with threading. I'm very new to asynchronous programming/threading except for knowing it's possible/doing very simple things with it, so I will probably make a separate discussion tab for going over my understanding of how Autopilot implements those things... been trying to get ffmpeg pipes to work for the evening lol

**Comment by adamsocrat** [comment]
## Is someone exposed github SMTP or what is this email?

Delivered-To: [my github email]
Received:  Wed, 28 Dec 2022 16:32:04 -0800 (PST)

ARC-Authentication-Results: i=1; mx.google.com;
       dkim=pass header.i=@github.com header.s=pf2014 
       spf=pass (google.com: domain of noreply@github.com designates 192.30.252.201 as permitted sender) smtp.mailfrom=noreply@github.com;
       dmarc=pass (p=REJECT sp=REJECT dis=NONE) header.from=github.com
Return-Path: <noreply@github.com>
Received: from smtp.github.com (out-18.smtp.github.com. [192.30.252.201])
        by mx.google.com with ESMTPS id m66-20020a672645000000b003b089fe31d4si2976947vsm.607.2022.12.28.16.32.04
        for <[my github email]>
        (version=TLS1_2 cipher=ECDHE-ECDSA-CHACHA20-POLY1305 bits=256/256);
        Wed, 28 Dec 2022 16:32:04 -0800 (PST)
Received-SPF: pass (google.com: domain of noreply@github.com designates 192.30.252.201 as permitted sender) client-ip=192.30.252.201;
Authentication-Results: mx.google.com;
       dkim=pass header.i=@github.com header.s=pf2014 header.b=gO1EbQAK;
       spf=pass (google.com: domain of noreply@github.com designates 192.30.252.201 as permitted sender) smtp.mailfrom=noreply@github.com;
       dmarc=pass (p=REJECT sp=REJECT dis=NONE) header.from=github.com
Received: from github-lowworker-f052de4.va3-iad.github.net (github-lowworker-f052de4.va3-iad.github.net [10.48.109.96]) by smtp.github.com (Postfix) with ESMTP id E53D13401F5 for <[my github email]>; Wed, 28 Dec 2022 16:32:03 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=githmy github email]ub.com; s=pf2014; t=1672273923; 
Date: Wed, 28 Dec 2022 16:32:03 -0800

**Reply by darkdvd** [reply]
I received the message 4 times...

**Reply by jmdelahanty** [reply]
Woah what in the world?? I only have my Salk email and my personal gmail with access to my GitHub account and don't see any suspicious logins or anything that I can see. I don't know how to tell if I've been compromised somehow or how to check properly.
